adding mitiq feature proposal template, fixes #498 (#548)
unitaryfund/mitiq: .github/ISSUE_TEMPLATE/feature_proposal.md

---
name: Feature proposal
about: Propose a great new feature for Mitiq
title: ''
labels: 'enhancement'
assignees: ''

---
## Pre-Request Checklist

> - [ ] I checked to make sure that this feature has not already been requested

Reviewer: ```suggestion
- [ ] I checked other issues on Mitiq to make sure that this feature has not already been requested
```

> - [ ] I or my organization would be willing to help implement and/or test this.

Reviewer: Should this be required? It seems reasonable for someone to request a feature but they can't work on it for whatever reason.

Author: I was just trying to emulate the format of other issue templates we have, see : https://github.com/unitaryfund/mitiq/blob/master/.github/ISSUE_TEMPLATE/bug_report.md 

It need not be required for us, I assume the volume of proposals we will get will not likely run into duplicates :)

> - [ ] This is **not** a breaking change to Mitiq.

Reviewer: I might remove this - some features could be breaking changes, no? and some quantum-but-not-software people may not be familiar with this terminology.

Author: I guess my thought here is not all need to be checked to be valid, its just more so we can gauge quickly how much work is involved. I take the point that breaking change might be not familiar outside of SW dev.

## Feature description

_TODO: Insert a short outline for what the proposed modification is._

> ## Considerations

Reviewer: Could we rename this section to "Purpose(s) of feature" or something similar? "Considerations"  is a bit vague to me.

Author: My intent is that it is broader than just purpose, in that it also asks about pros and cons of the approach outlined in the previous section and if there are other possible approaches to the problem the proposal. The purpose should have really be already described in the description of the feature. Maybe `##Discussion`` would be a better heading?

> _TODO: Explain why this modification to Mitiq is desirable. Briefly summarize the benefits and drawbacks of the chosen approach compared to other ways of achieving a similar functionality._

Reviewer: ```suggestion
_TODO: Explain the purpose of this feature and/or why this modification to Mitiq is desirable._
```

Reviewer: Could you clarify the second sentence? Does "other ways of achieving similar functionality" mean using other tools that already exist? Or alternative ways I thought of doing this in Mitiq?

Author: Exactly as you suggest, its asking about are there other ways to fulfill the purpose of the proposed feature, either with other tools, approaches, or just even different ways in Mitiq to go about implementing it. Sometimes there is only one clear way, but it can be helpful when discussing a new feature to check and see if any other approaches meet the user need without major changes.

> ## Context

Reviewer: "Context" also is a bit vague to me. What about (stealing from the description) "Important details" or "Important points going forward"?

Reviewer: Also, can we put this section after examples? I think examples are nice to see after the previous section describing the purpose and can elaborate on anything that perhaps wasn't clear there.

Author: Reordering is fine with me, and Context really does I think describe what I wanted here (which may not be a helpful thing i get). Basically the question here is are there other things we should consider like other papers or packages that implement part of the feature being proposed? Or is this similar to another technique that we may also want to consider implementing as it would be easy after adding this feature? Have they written up pseudo code for it already and have a link to that? It's the context that this proposal should be considered with. 

Again, maybe we want to know something different from the author and that's totally cool.
